### PR TITLE
Rpc incomplete blockchain

### PR DESCRIPTION
**Pull request by nacho692**
With the upcoming releases of Garbage collector #998 and State Importer #913. Some web3 responses require modifications.

This PR proposes this changes.

**eth_getBalance**
Lacking block: Nullpointer (_Internal Server Error_ **-32603**) -> BlockNotFound
Lacking state: 0 -> StateNotFound

**eth_getStorageAt**
Lacking block: Null -> BlockNotFound
Lacking state: 0 -> StateNotFound

**eth_getBlockTransactionCountByNumber**
Lacking block: Null -> BlockNotFound

**eth_getUncleCountByBlockHash**
Lacking block: Nullpointer (_Internal Server Error_ **-32603**) -> BlockNotFound

**eth_getUncleCountByBlockNumber**
Lacking block: Nullpointer (_Internal Server Error_ **-32603**) -> BlockNotFound

Rpc error code for both StateNotFound and BlockNotFound  is **-32600** which signifies an _Invalid Request_

:robot: *fed:rpc-incomplete-blockchain*

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/trie/MultiTrieStore.java [hunk 1]
```diff
@@ -73,9 +73,6 @@ public void flush() {
 
     /**
      * This method will go through all epochs from newest to oldest retrieving the <code>rootHash</code>
-     *
-     * @param rootHash the root of the {@link Trie} to retrieve
-     * @return an optional containing the {@link Trie} with <code>rootHash</code> if found
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/trie/MultiTrieStore.java]: These can go into the previous commit

File: rskj-core/src/main/java/co/rsk/trie/TrieStore.java [hunk 2]
```diff
@@ -25,7 +25,10 @@
 
     void flush();
 
-    /** Retrieve a trie node with the specified hash from this store */
+    /**
+     * @param hash the root of the {@link Trie} to retrieve
+     * @return an optional containing the {@link Trie} with <code>rootHash</code> if found
+     */
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/trie/TrieStore.java]: These can go into the previous commit

File: rskj-core/src/main/java/co/rsk/rpc/modules/rsk/RskModuleImpl.java [hunk 3]
```diff
@@ -143,7 +143,7 @@ private Block getByJsonBlockId(String id) {
         } else if ("latest".equalsIgnoreCase(id)) {
             return this.blockchain.getBestBlock();
         } else if ("pending".equalsIgnoreCase(id)) {
-            throw new JsonRpcUnimplementedMethodException("The method don't support 'pending' as a parameter yet");
+            throw RskJsonRpcRequestException.unimplemented("The method don't support 'pending' as a parameter yet");
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/rpc/modules/rsk/RskModuleImpl.java]: I'm not a big fan of this pattern, but I think we use to use _import static_ in this situations

File: rskj-core/src/main/java/org/ethereum/rpc/Web3Impl.java [hunk 4]
```diff
@@ -486,15 +485,15 @@ public String eth_getBlockTransactionCountByHash(String blockHash) throws Except
...
+    public static Block getBlockByNumberOrStr(String bnOrId, Blockchain blockchain) {
         Block b;
 
         if ("latest".equals(bnOrId)) {
             b = blockchain.getBestBlock();
         } else if ("earliest".equals(bnOrId)) {
             b = blockchain.getBlockByNumber(0);
         } else if ("pending".equals(bnOrId)) {
-            throw new JsonRpcUnimplementedMethodException("The method don't support 'pending' as a parameter yet");
+            throw RskJsonRpcRequestException.unimplemented("The method don't support 'pending' as a parameter yet");
```
Comment (diega) [inline rskj-core/src/main/java/org/ethereum/rpc/Web3Impl.java]: Prefer using _import static_

File: rskj-core/src/test/java/org/ethereum/rpc/Web3ImplScoringTest.java [hunk 5]
```diff
@@ -38,11 +38,9 @@
 import co.rsk.rpc.modules.txpool.TxPoolModuleImpl;
 import co.rsk.scoring.*;
 import co.rsk.test.World;
-import com.googlecode.jsonrpc4j.JsonRpcClientException;
 import org.bouncycastle.util.encoders.Hex;
 import org.ethereum.TestUtils;
 import org.ethereum.rpc.Simples.SimpleEthereum;
-import org.ethereum.rpc.exception.JsonRpcUnimplementedMethodException;
```
Comment (diega) [inline rskj-core/src/test/java/org/ethereum/rpc/Web3ImplScoringTest.java]: This changes belongs to the previous commit

File: rskj-core/src/main/java/co/rsk/rpc/Web3InformationRetriever.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    /**
+     * Retrieves an {@link AccountInformationProvider} based on the identifier.
+     * @param identifier {@link Web3InformationRetriever}
+     * @return The {@link AccountInformationProvider}
+     * @throws RskJsonRpcRequestException if the block indicated by the identifier is not found or the state for the
+     * block was not found.
+     */
+    public AccountInformationProvider getState(String identifier) {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/rpc/Web3InformationRetriever.java]: What about renaming `getState` to `getInformationProvider`? I think **state** is a huge concept in our domain

File: rskj-core/src/test/java/co/rsk/rpc/Web3InformationRetrieverTest.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.Assert.*;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class Web3InformationRetrieverTest {
```
Comment (diega) [inline rskj-core/src/test/java/co/rsk/rpc/Web3InformationRetrieverTest.java]: I think the _invalidParamError_ is never tested

File: rskj-core/src/main/java/org/ethereum/rpc/Web3Impl.java [hunk 8]
```diff
@@ -175,9 +173,9 @@ public void stop() {
         hashRateCalculator.stop();
     }
 
-    public int JSonHexToInt(String x) throws Exception {
+    private int JSonHexToInt(String x) {
         if (!x.startsWith("0x")) {
-            throw new Exception("Incorrect hex syntax");
+            throw RskJsonRpcRequestException.invalidParamError("Incorrect hex syntax");
```
Comment (diega) [inline rskj-core/src/main/java/org/ethereum/rpc/Web3Impl.java]: Prefer using _import static_

File: rskj-core/src/test/java/co/rsk/mine/TransactionModuleTest.java [hunk 9]
```diff
@@ -404,8 +405,8 @@ private Web3Impl createEnvironment(Blockchain blockchain, MiningMainchainView ma
                 null,
                 configCapabilities,
                 null,
-                null
-        );
+                null,
+                Mockito.mock(Web3InformationRetriever.class));
```
Comment (diega) [inline rskj-core/src/test/java/co/rsk/mine/TransactionModuleTest.java]: Passing `null` wouldn't work here?

File: rskj-core/src/test/java/co/rsk/rpc/Web3ImplRpcTest.java [hunk 10]
```diff
@@ -48,8 +50,8 @@ public void getRpcModules() {
                             null, null, null, null,
                             null, null, null, repositoryLocator, null,
                             null, null, null, null, null,
-                            null, null, null, null
-        );
+                            null, null, null, null,
+                mock(Web3InformationRetriever.class));
```
Comment (diega) [inline rskj-core/src/test/java/co/rsk/rpc/Web3ImplRpcTest.java]: Passing `null` wouldn't work here?

File: rskj-core/src/test/java/co/rsk/rpc/Web3RskImplTest.java [hunk 11]
```diff
@@ -111,25 +113,25 @@ public void web3_ext_dumpState() throws Exception {
                 null,
                 null,
                 null,
-                null
-        );
+                null,
+                mock(Web3InformationRetriever.class));
```
Comment (diega) [inline rskj-core/src/test/java/co/rsk/rpc/Web3RskImplTest.java]: Passing `null` wouldn't work here?

File: rskj-core/src/test/java/org/ethereum/rpc/Web3ImplScoringTest.java [hunk 12]
```diff
@@ -396,8 +399,8 @@ private static Web3Impl createWeb3(PeerScoringManager peerScoringManager) {
                 null,
                 null,
                 null,
-                null
-        );
+                null,
+                mock(Web3InformationRetriever.class));
```
Comment (diega) [inline rskj-core/src/test/java/org/ethereum/rpc/Web3ImplScoringTest.java]: Passing `null` wouldn't work here?

File: rskj-core/src/test/java/org/ethereum/rpc/Web3ImplSnapshotTest.java [hunk 13]
```diff
@@ -191,15 +193,14 @@ private Web3Impl createWeb3(SimpleEthereum ethereum) {
...
                 null,
                 null,
                 null,
                 null,
                 null,
                 null,
-                null
-        );
+                null,
+                mock(Web3InformationRetriever.class));
```
Comment (diega) [inline rskj-core/src/test/java/org/ethereum/rpc/Web3ImplSnapshotTest.java]: Passing `null` wouldn't work here?

File: rskj-core/src/test/java/org/ethereum/rpc/Web3ImplTest.java [hunk 14]
```diff
@@ -276,15 +275,14 @@ public void eth_mining()  {
...
                 null,
                 null,
                 null,
                 null,
                 null,
                 null,
-                null
-        );
+                null,
+                mock(Web3InformationRetriever.class));
```
Comment (diega) [inline rskj-core/src/test/java/org/ethereum/rpc/Web3ImplTest.java]: Passing `null` wouldn't work here?

File: rskj-core/src/test/java/co/rsk/mine/TransactionModuleTest.java (hunk 9 above)
Comment (nacho692) [reply rskj-core/src/test/java/co/rsk/mine/TransactionModuleTest.java]: I guess It would, personally I don't like sending null dependencies

**Pull request merged**